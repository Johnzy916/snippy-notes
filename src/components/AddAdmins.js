import React, { useEffect, useRef } from 'react'
import { connect } from 'react-redux'
import isEmail from 'validator/lib/isEmail'
import { useAlert } from 'react-alert'
import Dropdown from 'react-dropdown'
import { expandSection, collapseSection } from '../utils/expandSection'
import { startAddProjectAdmin, startRemoveProjectAdmin, setCurrentProject } from '../actions/admin'

export const AddAdmins = ({ addAdmin, removeAdmin, projectAdmins, isProjectAdmin,
                          isSuperAdmin, currentProject, setProject, allProjects }) => {
    
    // refs
    const areaRef = useRef()
    const chevronRef = useRef()

    // initiate react-alert
    const alert = useAlert()

    // collapse expandable area on load
    useEffect(() => {
      collapseSection(areaRef.current)
    }, [])

    // handle form submit
    const handleSubmit = (e) => {
      e.preventDefault();
      const adminEmail = e.target.children[1]

      if (!isEmail(adminEmail.value)) {
        alert.show('Please enter a valid email!', { type: 'error' })
      } else if (isSuperAdmin || isProjectAdmin) {
        if (projectAdmins[currentProject].includes(adminEmail.value)) {
          alert.show('This user is already a project admin!')
        } else {
          addAdmin(currentProject, adminEmail.value);
          adminEmail.value = '';
        }
      } else {
        alert.show('Only admins can add project admins!')
      }
    }

    // handle item remove
    const handleRemove = (e) => {
      const email = e.target.previousElementSibling.textContent

      if (isSuperAdmin || isProjectAdmin) {
        if (projectAdmins[currentProject].length === 1) {
          alert.show('There must be at least one project admin!')
        } else {
          removeAdmin(currentProject, email)
        }
      } else {
        alert.show('You must be an admin to do this!')
      }
    }

    // handle section expansion
    const handleExpansion = () => {
      const area = areaRef.current
      const chevron = chevronRef.current
      const isCollapsed = area.getAttribute('data-collapsed') === 'true';
        
      if(isCollapsed) {
        expandSection(area)
        area.setAttribute('data-collapsed', 'false')
        chevron.style.transform = 'rotate(180deg)';
      } else {
        collapseSection(area)
        chevron.style.transform = 'rotate(0deg)';
      }
    }

  return (
    <div className="admin-module">
      <div className="admin-section__header">
        Add Project Admin
      </div>
      <div className="admin-section">
      <div className="add-admin">
          <form className="add-admin__input-form"
            id="add-admin-form"
            onSubmit={handleSubmit}>
            {
              (allProjects && currentProject) && (
                allProjects.length < 2 ? (
                  <div className="add-admin__project-name">
                    { 
                      currentProject.charAt(0).toUpperCase() + currentProject.slice(1)
                    }
                  </div>
                ) : (
                  <Dropdown 
                    options={allProjects.map(project => project.charAt(0).toUpperCase() + project.slice(1))}
                    value={currentProject.charAt(0).toUpperCase() + currentProject.slice(1)}
                    onChange={(e) => setProject(e.value.toLowerCase())} />
                )
              )
            }
            <input className="add-admin__input add-admin__email"
                placeholder="Google account email"
                contentEditable >
            </input>
            <button className="btn btn--secondary btn--shine btn--add"
              type="submit">
                &#x2b;
            </button>
          </form>
        </div>
      </div>
      <div className="admin-expandable-area-container expandable-app-admin-area"
        ref={areaRef}>
        <div className="admin-expandable-area">
            {
              (projectAdmins && currentProject) && (
                projectAdmins[currentProject].sort((a, b) => a < b ? -1 : 1).map(email => {
                  return (
                    <div className="admin-area__item" key={email}>
                        <div className="admin-area__item-text">
                          { email }
                        </div>
                        <button className="admin-area__item-remove"
                        onClick={handleRemove}>
                          &#x2715;
                        </button>
                    </div>
                  )
                })
              )
            }
        </div>
      </div>
      <div className="admin-section__footer"
        onClick={handleExpansion}>
        View admins
        <div className="chevron-logo-container admin-chevron">
          <svg ref={chevronRef} xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"/></svg>
        </div>
      </div>
    </div>
  );
}

const mapStateToProps = (state) => ({
  projectAdmins: state.admin.admins,
  isSuperAdmin: state.admin.role === 'superAdmin',
  isProjectAdmin: state.admin.role === 'projectAdmin',
  currentProject: state.admin.currentProject,
  allProjects: state.admin.allProjects || []
})

const mapDispatchToProps = (dispatch) => ({
  addAdmin: (project, email) => dispatch(startAddProjectAdmin(project, email)),
  removeAdmin: (project, email) => dispatch(startRemoveProjectAdmin(project, email)),
  setProject: (project) => dispatch(setCurrentProject(project)),
})

export default connect(mapStateToProps, mapDispatchToProps)(AddAdmins);