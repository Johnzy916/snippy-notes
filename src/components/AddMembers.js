import React, { useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import Dropdown from 'react-dropdown';
import isEmail from 'validator/lib/isEmail';
import { useAlert } from 'react-alert';
import { expandSection, collapseSection } from '../utils/expandSection';
import { startAddTeamMember, startRemoveTeamMember, setCurrentTeam, startRemoveTeamAdmin } from '../actions/admin';

export const AddMembers = ({ isTeamAdmin, currentTeam, addMember, setTeam, isDataLoaded,
                              teamMembers, removeMember, teamAdmins, isProjectAdmin,
                              userEmail, isSuperAdmin, removeAdmin }) => {

    // refs
    const areaRef = useRef()
    const chevronRef = useRef()

    // initiate react-alert
    const alert = useAlert()

    // collapse expandable area
    useEffect(() => {
      collapseSection(areaRef.current)
    }, [])

    // handle form submit
    const handleSubmit = (e) => {
      e.preventDefault();
      const memberEmail = e.target.children[1]
      if (!isEmail(memberEmail.value)) {
        alert.show('Please enter a valid email!', { type: 'error' })
      } else if (teamMembers[currentTeam].includes(memberEmail.value)) {
        alert.show('User is already a member of the team!')
      } else if (isSuperAdmin || isProjectAdmin || isTeamAdmin.includes(currentTeam)) {
        addMember(currentTeam, memberEmail.value);
        memberEmail.value = '';
      } else {
        alert.show('You must be an admin to add team members!')
      }
    }

    // handle item remove
    const handleRemove = (e) => {
      const member = e.target.previousElementSibling.textContent
      if (isProjectAdmin || isSuperAdmin ||
        !e.target.closest('div').classList.contains('admin-area__item--admin') ||
        member === userEmail) {
          const userIsAdmin = teamAdmins[currentTeam].includes(member)
          // if there's only one admin left, don't remove them
          if (userIsAdmin && Object.keys(teamAdmins[currentTeam]).length === 1) {
            alert.show(`There should be at least one team admin!`)
          //  else if member is an admin, remove the admin and member
          } else if (userIsAdmin) {
            removeAdmin(currentTeam, member) // removes both
            // else, just remove the member
          } else {
            removeMember(currentTeam, member)
          }
      } else {
        alert.show(`You can't remove another admin!`)
      }
    }

    // handle section expansion
    const handleExpansion = () => {
      const area = areaRef.current
      const chevron = chevronRef.current
      const isCollapsed = area.getAttribute('data-collapsed') === 'true';
        
      if(isCollapsed) {
        expandSection(area)
        area.setAttribute('data-collapsed', 'false')
        chevron.style.transform = 'rotate(180deg)';
      } else {
        collapseSection(area)
        chevron.style.transform = 'rotate(0deg)';
      }
    }


  return (
    <div className="admin-module">
      <div className="admin-section__header">
        Add Team Members
      </div>
      <div className="admin-section">
      <div className="add-member">
          <form className="add-member__input-form"
            id="add-member-form"
            onSubmit={handleSubmit}>
            {
              (teamMembers && currentTeam) && (
                Object.keys(teamMembers).length < 2 ? (
                  <div className="add-member__team-name">
                    { 
                      Object.keys(teamMembers)[0].charAt(0).toUpperCase() + Object.keys(teamMembers)[0].slice(1)
                    }
                  </div>
                  ) : (
                  <Dropdown 
                    options={Object.keys(teamMembers)
                              .map(team => team.charAt(0).toUpperCase() + team.slice(1))}
                    value={currentTeam.charAt(0).toUpperCase() + currentTeam.slice(1)}
                    onChange={(e) => setTeam(e.value.toLowerCase())} />
                )
              )
            }
            <input className="add-member__input add-member__email"
                placeholder="Google account email"
                contentEditable >
            </input>
            <button className="btn btn--secondary btn--shine btn--add"
              type="submit">
              &#x2b;
            </button>
          </form>
        </div>
      </div>
      <div className="admin-expandable-area-container expandable-member-area"
        ref={areaRef}>
        <div className="admin-expandable-area__notification--white">
            <div className="admin__key-square"></div> = Team admin
        </div>
        <div className="admin-expandable-area">
                {
                  isDataLoaded && (
                    teamMembers[currentTeam].sort((a, b) => a < b ? -1 : 1).map(member => {
                      return (
                        // add dom class for admins to highlight in ui
                        <div className={ teamAdmins[currentTeam].includes(member) ?
                                        "admin-area__item--admin"
                                      : "admin-area__item"} key={member}>
                            <div className="admin-area__item-text">
                              { member }
                            </div>
                            <button className="admin-area__item-remove"
                            onClick={handleRemove}>
                              &#x2715;
                            </button>
                        </div>
                      );
                    })
                  )
                }
        </div>
      </div>
      <div className="admin-section__footer"
        onClick={handleExpansion}>
        View members
        <div className="chevron-logo-container member-chevron">
          <svg ref={chevronRef} xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"/></svg>
        </div>
      </div>
    </div>
  );
}

const mapStateToProps = (state) => ({
  isTeamAdmin: state.admin.isTeamAdmin,
  currentTeam: state.admin.currentTeam,
  teamMembers: state.admin.teamMembers || {},
  teamAdmins: state.admin.teamAdmins,
  userEmail: state.user.email,
  isProjectAdmin: state.admin.role === 'projectAdmin',
  isSuperAdmin: state.admin.role === 'superAdmin',
  isDataLoaded: state.admin.currentTeam && state.admin.teamMembers && 
                state.admin.teamAdmins
})

const mapDispatchToProps = (dispatch) => ({
  addMember: (team, email) => dispatch(startAddTeamMember(team, email)),
  setTeam: (team) => dispatch(setCurrentTeam(team)),
  removeMember: (team, email) => dispatch(startRemoveTeamMember(team, email)),
  removeAdmin: (team, email) => dispatch(startRemoveTeamAdmin(team, email))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddMembers);