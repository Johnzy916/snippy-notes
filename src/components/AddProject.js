import React, { useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import isEmail from 'validator/lib/isEmail';
import { useAlert } from 'react-alert';
import { startAddProject, startRemoveProject } from '../actions/admin';
import { expandSection, collapseSection } from '../utils/expandSection';

export const AddProject = ({ addProject, allProjects, removeProject,
                            isSuperAdmin }) => {

  // refs                            
  const areaRef = useRef()
  const chevronRef = useRef()

  // initiate react-alert
  const alert = useAlert()

  // collapse expandable area
  useEffect(() => {
    collapseSection(areaRef.current)
  }, [])

  // handle form submit
  const handleSubmit = (e) => {
    e.preventDefault();
    const nameInput = e.target.children[0]
    const adminInput = e.target.children[1]
    if (!nameInput.value || !adminInput.value) {
      alert.show('Both inputs are required!')
    } else if (!isEmail(adminInput.value)) {
      alert.show('Please enter a valid email!', { type: 'error' })
    } else if (isSuperAdmin) {
      addProject(nameInput.value, adminInput.value);
      nameInput.value = '';
      adminInput.value = '';
      nameInput.focus();
      // if project already exists,
      // show message that new admin was added
      if (allProjects.includes(nameInput.value)) {
        alert.show('Project already exists. New admin added!')
      }
    } else {
      alert.show('You must be a super admin to do this!')
    }
  }

  // handle remove item
  const handleRemove = (e) => {
    const project = e.target.previousElementSibling.textContent
    if (isSuperAdmin) {
    removeProject(project)
    } else {
      alert.show('You must be a super admin to do this!')
    }
  }

  // handle section expansion
  const handleExpansion = () => {
    const area = areaRef.current
    const chevron = chevronRef.current
    const isCollapsed = area.getAttribute('data-collapsed') === 'true';
      
    if(isCollapsed) {
      expandSection(area)
      area.setAttribute('data-collapsed', 'false')
      chevron.style.transform = 'rotate(180deg)';
    } else {
      collapseSection(area)
      chevron.style.transform = 'rotate(0deg)';
    }
  }

  return (
    <div className="admin-module">
      <div className="admin-section__header">
        Add New Project
      </div>
      <div className="admin-section">
        <div className="add-team">
            <form className="add-team__input-form"
              id="add-project-form"
              onSubmit={handleSubmit}>
              <input className="add-team__input add-project__name"
                  placeholder="Project name"
                  contentEditable >
              </input>
              <input className="add-team__input add-project__admin"
                  placeholder="Project admin (Google email)"
                  contentEditable>
              </input>
              <button className="btn btn--secondary btn--shine btn--add"
                type="submit"
                form="add-project-form">
                &#x2b;
              </button>
            </form>
        </div>
      </div>
      <div className="admin-expandable-area-container expandable-project-area"
        ref={areaRef}>
        <div className="admin-expandable-area__notification">
            Deleting a project will remove ALL of it's teams and data!
        </div>
        <div className="admin-expandable-area">
        {
          allProjects && (
            allProjects.length > 0 ? (
              allProjects.sort((a, b) => a > b ? 1 : -1).map(project => {
                return (
                  <div className="admin-area__item" key={project}>
                      <div className="admin-area__item-text">
                        { project.charAt(0).toUpperCase() + project.slice(1) }
                      </div>
                      <button className="admin-area__item-remove"
                      onClick={handleRemove}>
                        &#x2715;
                      </button>
                  </div>
                )
              })
            ) : (
              <div className="list-item__message">
                  There are no projects. Add the first one!
              </div>
            )
          )
        }
        </div>
      </div>
      <div className="admin-section__footer"
        onClick={handleExpansion}>
        View projects 
        <div className="chevron-logo-container project-chevron">
          <svg ref={chevronRef} xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"/></svg>
        </div>
      </div>
    </div>
  );
}

const mapStateToProps = (state) => ({
  allProjects: state.admin.allProjects,
  isSuperAdmin: state.admin.role === 'superAdmin',
})

const mapDispatchToProps = (dispatch) => ({
  addProject: (name, admin) => dispatch(startAddProject(name, admin)),
  removeProject: (project) => dispatch(startRemoveProject(project))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddProject);