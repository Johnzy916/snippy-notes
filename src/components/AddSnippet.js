import React, { useEffect, useRef } from 'react'
import { connect } from 'react-redux'
import DOMPurify from 'dompurify'
import { useAlert } from 'react-alert'
import isEmpty from 'validator/lib/isEmpty'
import { startAddUserSnippet } from '../actions/user'
import { startAddTeamSnippet } from '../actions/teams'
import { updateUserAddSnippet, updateTeamAddSnippets } from '../actions/temp'
import { copyText, pasteText } from '../utils/clipboard'

export const AddSnippet = ({ addUserSnippet, addTeamSnippet, type, team,
                            userSnippets, teamSnippets, userAddSnippet,
                            teamAddSnippets, updateUserSnippet, updateTeamSnippets }) => {

  // refs
  const shortcutRef = useRef()
  const textRef = useRef()

  // initiate react-alert
  const alert = useAlert();

  // save and insert on mount or unmount
  useEffect(() => {
      if (team && teamAddSnippets && teamAddSnippets[team]) {
        textRef.current.innerHTML = DOMPurify.sanitize(teamAddSnippets[team].text)
      } else if (!team && userAddSnippet) {
        textRef.current.innerHTML = DOMPurify.sanitize(userAddSnippet.text)
      }

      return () => {
          const snippet = { 
            shortcut: shortcutRef.current.textContent,
            text: DOMPurify.sanitize(textRef.current.innerHTML)
          }
          if (team) {
            updateTeamSnippets(team, snippet)
          } else {
            updateUserSnippet(snippet)
          }
      }
  }, [])

  // handle form submit
  const handleSubmit = (e) => {
    e.preventDefault()
    const shortcut = e.target.children[0]
    const text = e.target.children[1]
    if (isEmpty(shortcut.textContent) || isEmpty(text.innerHTML)) {
      alert.show('Please fill out both inputs!')
    // check if shortcut already exists in personal snippets
    } else if (userSnippets.some(el => el.shortcut === shortcut.textContent)) {
      alert.show('This key is already used in your snippets!')
    // check if shortcut already exists in other team snippets
    } else if ([].concat(...Object.values(teamSnippets)
                  .map(arr => arr.map(obj => obj.shortcut)))
                  .includes(shortcut.textContent)) {
     alert.show('This key is already used in your team snippets!')
    } else {
      // only admins can add snippets, otherwise it's the user's page
      (team && type === 'admin') ?
        addTeamSnippet(team, { 
          shortcut: shortcut.textContent,
          text: DOMPurify.sanitize(text.innerHTML)
        })
      : addUserSnippet({ 
          shortcut: shortcut.textContent, 
          text: DOMPurify.sanitize(text.innerHTML)
        })
      shortcut.textContent = '';
      text.textContent = '';
      shortcut.focus()
    }
  }

  return (
    <form className="add-snippet"
      id="add-snippet-form"
      onSubmit={handleSubmit} >
      <div className="add-snippet__input add-snippet__shortcut"
        data-placeholder="Key"
        ref={shortcutRef}
        data-team={team}
        contentEditable
        suppressContentEditableWarning
        onCopy={(e) => {
          e.preventDefault()
          copyText(e)
        }}
        onPaste={(e) => {
          e.preventDefault()
          pasteText(e)
        }}>
          {
            (team && teamAddSnippets && teamAddSnippets[team]) ?
              teamAddSnippets[team].shortcut
            : (!team && userAddSnippet) ?
              userAddSnippet.shortcut
            : ''
          }
      </div>
      <div className="add-snippet__input add-snippet__text"
        data-placeholder="Snippet"
        ref={textRef}
        data-team={team}
        contentEditable
        onCopy={(e) => {
          e.preventDefault()
          copyText(e)
        }}
        onPaste={(e) => {
          e.preventDefault()
          pasteText(e)
        }}>

      </div>
      <button className="btn btn--secondary btn--shine btn--add"
        type="submit" >
        &#x2b;
      </button>
    </form>
  );
}

const mapStateToProps = (state) => ({
  userSnippets: state.user.snippets || [],
  teamSnippets: state.teams.snippets || [],
  userAddSnippet: state.temp.user,
  teamAddSnippets: state.temp.teams
});

const mapDispatchToProps = (dispatch) => ({
  addUserSnippet: (snippetData) => dispatch(startAddUserSnippet(snippetData)),
  addTeamSnippet: (team, snippetData) => dispatch(startAddTeamSnippet(team, snippetData)),
  updateUserSnippet: (snippet) => dispatch(updateUserAddSnippet(snippet)),
  updateTeamSnippets: (team, snippet) => dispatch(updateTeamAddSnippets(team, snippet))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddSnippet)