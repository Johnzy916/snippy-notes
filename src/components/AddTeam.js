import React, { useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import isEmail from 'validator/lib/isEmail';
import { useAlert } from 'react-alert';
import { startAddTeam, startRemoveTeam } from '../actions/admin';
import { expandSection, collapseSection } from '../utils/expandSection';

export const AddTeam = ({ addTeam, allTeams, removeTeam, isSuperAdmin,
                          isProjectAdmin, currentProject }) => {

  // refs
  const areaRef = useRef()
  const chevronRef = useRef()

  // initiate react-alert
  const alert = useAlert()

  // collapse expandable area
  useEffect(() => {
    collapseSection(areaRef.current)
  }, [])

  // handle form submit
  const handleSubmit = (e) => {
    e.preventDefault();
    const nameInput = e.target.children[0];
    const adminInput = e.target.children[1];
    if (!nameInput.value || !adminInput.value) {
      alert.show('Both inputs are required!')
    } else if (!isEmail(adminInput.value)) {
      alert.show('Please enter a valid email!', { type: 'error' })
    } else if (isSuperAdmin || isProjectAdmin) {
      addTeam(currentProject, nameInput.value, adminInput.value);
      nameInput.value = '';
      adminInput.value = '';
      nameInput.focus();
      // if team already exists,
      // show message that new admin was added
      if (allTeams.includes(nameInput.value)) {
        alert.show('Team already exists. New admin added!')
      }
    } else {
      alert.show('You must be an admin to add teams!')
    }
  }

  // handle item remove
  const handleRemove = (e) => {
    const team = e.target.previousElementSibling.textContent
    if (isSuperAdmin || isProjectAdmin) {
      removeTeam(currentProject, team)
    } else {
      alert.show('You must be an admin to remove teams!')
    }
  }

  // handle section expansion
  const handleExpansion = () => {
    const area = areaRef.current
    const chevron = chevronRef.current
    const isCollapsed = area.getAttribute('data-collapsed') === 'true';
      
    if(isCollapsed) {
      expandSection(area)
      area.setAttribute('data-collapsed', 'false')
      chevron.style.transform = 'rotate(180deg)';
    } else {
      collapseSection(area)
      chevron.style.transform = 'rotate(0deg)';
    }
  }

  return (
    <div className="admin-module">
      <div className="admin-section__header">
        Add New Team
      </div>
      <div className="admin-section">
        <div className="add-team">
            <form className="add-team__input-form"
              id="add-team-form"
              onSubmit={handleSubmit}>
              <input className="add-team__input add-team__name"
                  placeholder="Team name"
                  contentEditable >
              </input>
              <input className="add-team__input add-team__admin"
                  placeholder="Team admin (Google email)"
                  contentEditable>
              </input>
              <button className="btn btn--secondary btn--shine btn--add"
                type="submit"
                form="add-team-form">
                &#x2b;
              </button>
            </form>
        </div>
      </div>
      <div className="admin-expandable-area-container expandable-team-area"
        ref={areaRef}>
        <div className="admin-expandable-area__notification">
            Deleting a team will remove their snippets!
        </div>
        <div className="admin-expandable-area">
        {
          allTeams && (
            allTeams.length > 0 ? (
              allTeams.sort((a, b) => a > b ? 1 : -1).map(team => {
                return (
                  <div className="admin-area__item" key={team}>
                      <div className="admin-area__item-text">
                        { team.charAt(0).toUpperCase() + team.slice(1) }
                      </div>
                      <button className="admin-area__item-remove"
                      onClick={handleRemove}>
                        &#x2715;
                      </button>
                  </div>
                )
              })
            ) : (
              <div className="list-item__message">
                  There are no teams for {currentProject ? currentProject.charAt(0).toUpperCase() + currentProject.slice(1) : `this project`}. Add the first one!
              </div>
            )
          )
        }
        </div>
      </div>
      <div className="admin-section__footer"
        onClick={handleExpansion}>
        View teams 
        <div className="chevron-logo-container team-chevron">
          <svg ref={chevronRef} xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"/></svg>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => ({
  allTeams: state.admin.allTeams || [],
  isSuperAdmin: state.admin.role === 'superAdmin',
  isProjectAdmin: state.admin.role === 'projectAdmin',
  currentProject: state.admin.currentProject,
})

const mapDispatchToProps = (dispatch) => ({
  addTeam: (project, name, admin) => dispatch(startAddTeam(project, name, admin)),
  removeTeam: (project, team) => dispatch(startRemoveTeam(project, team))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddTeam);