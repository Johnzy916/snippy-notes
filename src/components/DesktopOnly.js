import React from 'react';

const DesktopOnly = () => {
    // overlay for when the screen is less than 550px
    // used for MVP as app is intended for desktop
    // would need to cover tracking keypress for Android and others
    // and tighten up media query design for mobile
    // -- can be found in PrivateRoute and AdminRoute
    return (
        <div className="mobile-not-available">
            This app was built for use on a desktop!
            <div className="desktop-overlay-img"></div>
            If you're on a desktop, please widen the screen.
        </div>
    );
}

export default DesktopOnly;