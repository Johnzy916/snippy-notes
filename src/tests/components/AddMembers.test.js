import React from 'react'
import { shallow } from 'enzyme'
import { AddMembers } from '../../components/AddMembers'
import data from '../fixtures/adminData'

const state = {
    ...data,
    isDataLoaded: true,
    isSuperAdmin: true
}

let wrapper, addMember, setTeam, removeMember, removeAdmin

beforeEach(() => {
    addMember = jest.fn()
    setTeam = jest.fn()
    removeMember = jest.fn()
    removeAdmin = jest.fn()
    wrapper = shallow(<AddMembers {...state}
                    addMember={addMember}
                    setTeam={setTeam}
                    removeMember={removeMember}
                    removeAdmin={removeAdmin}
                    />)
})

it('should render AddMembers with no dropdown', () => {
    const props = {
        ...state,
        teamMembers: { team: state.teamMembers.team }
    }
    wrapper = shallow(<AddMembers {...props} />)
    expect(wrapper).toMatchSnapshot()
})

it('should render AddMembers with dropdown', () => {
    expect(wrapper).toMatchSnapshot()
})

it('should render team member emails with correct class', () => {
    expect(wrapper.find('.admin-area__item--admin')).toHaveLength(1)
    expect(wrapper.find('.admin-area__item')).toHaveLength(1)
})

it('should call setTeam on Dropdown change', () => {
    const value = state.allTeams[1]
    wrapper.find('Dropdown').prop('onChange')({ value })
    expect(setTeam).toHaveBeenLastCalledWith(value)
})

it('should call addMember on valid input (as super/project admin)', () => {
    const value = 'test@email.com'
    const target = { children: [ '', { value } ]}
    wrapper.find('#add-member-form').simulate('submit', {
        preventDefault: () => {},
        target
    })
    expect(addMember).toHaveBeenLastCalledWith(state.currentTeam, value)
})

it('should call addMember on valid input (as teamAdmin)', () => {
    const value = 'test@email.com'
    const target = { children: [ '', { value } ]}
    const props = {
        ...state, isSuperAdmin: false
    }
    wrapper = shallow(<AddMembers {...props} addMember={addMember} />)
    wrapper.find('#add-member-form').simulate('submit', {
        preventDefault: () => {},
        target
    })
    expect(addMember).toHaveBeenLastCalledWith(state.currentTeam, value)
})

it('should call removeAdmin when admin is removed', () => {
    const email = state.teamMembers[state.currentTeam][0]
    wrapper.find(`.admin-area__item-remove`).at(0).simulate('click', {
        target: { previousElementSibling: { textContent: email }}
    })
    expect(removeAdmin).toHaveBeenLastCalledWith(state.currentTeam, email)
})

it('should call removeMember when member is removed', () => {
    const email = state.teamMembers[state.currentTeam][1]
    wrapper.find(`.admin-area__item-remove`).at(1).simulate('click', {
        target: { previousElementSibling: { textContent: email }}
    })
    expect(removeMember).toHaveBeenLastCalledWith(state.currentTeam, email)
})