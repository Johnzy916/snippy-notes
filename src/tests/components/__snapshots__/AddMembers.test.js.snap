// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddMembers with dropdown 1`] = `
<div
  className="admin-module"
>
  <div
    className="admin-section__header"
  >
    Add Team Members
  </div>
  <div
    className="admin-section"
  >
    <div
      className="add-member"
    >
      <form
        className="add-member__input-form"
        id="add-member-form"
        onSubmit={[Function]}
      >
        <Dropdown
          baseClassName="Dropdown"
          onChange={[Function]}
          options={
            Array [
              "Team",
              "Team2",
            ]
          }
          value="Team"
        />
        <input
          className="add-member__input add-member__email"
          contentEditable={true}
          placeholder="Google account email"
        />
        <button
          className="btn btn--secondary btn--shine btn--add"
          type="submit"
        >
          +
        </button>
      </form>
    </div>
  </div>
  <div
    className="admin-expandable-area-container expandable-member-area"
  >
    <div
      className="admin-expandable-area__notification--white"
    >
      <div
        className="admin__key-square"
      />
       = Team admin
    </div>
    <div
      className="admin-expandable-area"
    >
      <div
        className="admin-area__item--admin"
        key="email"
      >
        <div
          className="admin-area__item-text"
        >
          email
        </div>
        <button
          className="admin-area__item-remove"
          onClick={[Function]}
        >
          ✕
        </button>
      </div>
      <div
        className="admin-area__item"
        key="email5"
      >
        <div
          className="admin-area__item-text"
        >
          email5
        </div>
        <button
          className="admin-area__item-remove"
          onClick={[Function]}
        >
          ✕
        </button>
      </div>
    </div>
  </div>
  <div
    className="admin-section__footer"
    onClick={[Function]}
  >
    View members
    <div
      className="chevron-logo-container member-chevron"
    >
      <svg
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`should render AddMembers with no dropdown 1`] = `
<div
  className="admin-module"
>
  <div
    className="admin-section__header"
  >
    Add Team Members
  </div>
  <div
    className="admin-section"
  >
    <div
      className="add-member"
    >
      <form
        className="add-member__input-form"
        id="add-member-form"
        onSubmit={[Function]}
      >
        <div
          className="add-member__team-name"
        >
          Team
        </div>
        <input
          className="add-member__input add-member__email"
          contentEditable={true}
          placeholder="Google account email"
        />
        <button
          className="btn btn--secondary btn--shine btn--add"
          type="submit"
        >
          +
        </button>
      </form>
    </div>
  </div>
  <div
    className="admin-expandable-area-container expandable-member-area"
  >
    <div
      className="admin-expandable-area__notification--white"
    >
      <div
        className="admin__key-square"
      />
       = Team admin
    </div>
    <div
      className="admin-expandable-area"
    >
      <div
        className="admin-area__item--admin"
        key="email"
      >
        <div
          className="admin-area__item-text"
        >
          email
        </div>
        <button
          className="admin-area__item-remove"
          onClick={[Function]}
        >
          ✕
        </button>
      </div>
      <div
        className="admin-area__item"
        key="email5"
      >
        <div
          className="admin-area__item-text"
        >
          email5
        </div>
        <button
          className="admin-area__item-remove"
          onClick={[Function]}
        >
          ✕
        </button>
      </div>
    </div>
  </div>
  <div
    className="admin-section__footer"
    onClick={[Function]}
  >
    View members
    <div
      className="chevron-logo-container member-chevron"
    >
      <svg
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </div>
  </div>
</div>
`;
